pkg/app/pipedv1/livestatereporter/livestatereporter_test.go

```diff
@@ -0,0 +1,294 @@
+	return &deployment.DetermineVersionsResponse{
+		Versions: []*model.ArtifactVersion{},
+	}, nil
+}
+func (p *fakePlugin) FetchDefinedStages(ctx context.Context, req *deployment.FetchDefinedStagesRequest, opts ...grpc.CallOption) (*deployment.FetchDefinedStagesResponse, error) {
+	stages := make([]string, 0, len(p.quickStages)+len(p.pipelineStages)+len(p.rollbackStages))
+
+	for _, s := range p.quickStages {
+		stages = append(stages, s.Name)
+	}
+	for _, s := range p.pipelineStages {
+		stages = append(stages, s.Name)
+	}
+	for _, s := range p.rollbackStages {
+		stages = append(stages, s.Name)
+	}
+	return &deployment.FetchDefinedStagesResponse{
+		Stages: stages,
+	}, nil
+}
+func (p *fakePlugin) GetLivestate(ctx context.Context, in *livestate.GetLivestateRequest, opts ...grpc.CallOption) (*livestate.GetLivestateResponse, error) {
+	return &livestate.GetLivestateResponse{
+		ApplicationLiveState: &model.ApplicationLiveState{},
+		SyncState:            &model.ApplicationSyncState{},
+	}, nil
+}
+
+type fakeAPILister struct {
+	applicationLister
+	apps []*model.Application
+}
+
+func (f *fakeAPILister) List() []*model.Application {
+	return f.apps
+}
+
+func Test_reporter_flushSnapshots(t *testing.T) {
+	gitClient, err := git.NewClient()
+	require.NoError(t, err)
+
+	pr := &reporter{
+		snapshotFlushInterval: 1 * time.Minute,
+		appLister: &fakeAPILister{
+			apps: []*model.Application{
+				&model.Application{
+					Id:   "app-id",
+					Name: "app-name",
+					GitPath: &model.ApplicationGitPath{
+						Repo: &model.ApplicationGitRepository{
+							Id:     "repo-id",
+							Remote: "https://github.com/pipe-cd/examples.git",
+							Branch: "master",
+						},
+						Path:           "kubernetes/canary",
+						ConfigFilename: "app.pipecd.yaml",
+					},
+				},
+			},
+		},
+		apiClient: &fakeAPIClient{},
+		repoMap:   make(map[string]git.Repo, 0),
+		pluginRegistry: func() plugin.PluginRegistry {
+			r, err := plugin.NewPluginRegistry(
+				context.Background(),
+				[]plugin.Plugin{
+					{
+						Name: "k8s",
+						Cli: &fakePlugin{
+							pipelineStages: []*model.PipelineStage{
+								{
+									Name: "K8S_CANARY_ROLLOUT",
+								},
+								{
+									Name: "K8S_CANARY_CLEAN",
+								},
+								{
+									Name: "K8S_PRIMARY_ROLLOUT",
+								},
+							},
+						},
+					},
+					{
+						Name: "wait",
+						Cli: &fakePlugin{
+							pipelineStages: []*model.PipelineStage{
+								{
+									Name: "WAIT",
+								},
+							},
+						},
+					},
+				},
+			)
+
+			require.NoError(t, err)
+			return r
+		}(),
+		gitClient: gitClient,
+		pipedConfig: &config.PipedSpec{
+			Repositories: []config.PipedRepository{
+				{
+					RepoID: "repo-id",
+					Remote: "https://github.com/pipe-cd/examples.git",
+					Branch: "master",
+				},
+			},
+		},
+		secretDecrypter: nil,
+		workingDir:      t.TempDir(),
+		logger:          zaptest.NewLogger(t),
+	}
+
+	pr.flushSnapshots(context.Background())
```

Comment:
There are no assertion statements.
Does this test only assert that no panic occurred? If so, it's better to comment on that.

Reply:
Thanks. I added a comment and a new test for `flush` to check the logic. (currently added only the success case)
[8b8f4c6](https://github.com/pipe-cd/pipecd/pull/5457/commits/8b8f4c6a5ced90073a91c8c29fbe8ace34cc581b)